lsf/man/man1/lsload.1
.ds ]W %
.ds ]L
.nh
.TH lsload 1 "OpenLava Version 3.3 - Mar 2016"
.br
.SH NAME
\fBlsload\fR - displays load information for hosts
.SH SYNOPSIS
.BR
.PP
.PP
\fBlsload\fR [\fB-l\fR] [\fB-N\fR | \fB-E\fR] [\fB-I\fR\fB \fR\fIload_index\fR[\fB:\fR\fIload_index\fR] \fI...\fR] [\fB-n\fR\fB \fR\fInum_hosts\fR] 
[\fB-R\fR\fB \fR\fIres_req\fR] \fIhost_name\fR\fI ...  ...\fR
.PP
\fBlsload\fR \fB-s\fR\fB \fR[\fIresource_name\fR\fI ...\fR]
.PP
\fBlsload\fR [\fB-h\fR | \fB-V\fR]
.SH DESCRIPTION
.BR
.PP
.PP
\fB\fRDisplays load information for hosts. Load information can be displayed 
on a per-host basis, or on a per-resource basis.
.PP
By default, displays load information for all hosts in the local cluster, 
per host. 
.PP
By default, displays raw load indices.
.PP
By default, load information for resources is displayed according to CPU 
and paging load.
.SH OPTIONS
.BR
.PP
.TP 
\fB-l
\fR
.IP
Long format. Displays load information without truncation along with 
additional fields for I/O and external load indices.

.IP


.IP
This option overrides the index names specified with the -I option.


.TP 
\fB-N
\fR
.IP
Displays normalized CPU run queue length load indices (see 
lsfintro(1)).


.TP 
\fB-E
\fR
.IP
Displays effective CPU run queue length load indices (see 
lsfintro(1)). Options -N and -E are mutually exclusive.


.TP 
\fB-I\fR \fIload_index\fR[\fB:\fR\fIload_index\fR] ...

.IP
Displays only load information for the specified load indices. Load 
index names must be separated by a colon (for example, r1m:pg:ut).


.TP 
\fB-n\fR\fI \fR\fInum_hosts
\fR
.IP
Displays only load information for the requested number of hosts. 
Information for up to \fInum_hosts\fR hosts that best satisfy the resource 
requirements is displayed.


.TP 
\fB-R\fR\fI \fR\fIres_req
\fR
.IP
Displays only load information for hosts that satisfy the specified 
resource requirements. See lsinfo(1) for a list of built-in resource 
names.

.IP
Load information for the hosts is sorted according to load on the 
specified resources.

.IP
If \fIres_req\fR contains special resource names, only load information for 
hosts that provide these resources is displayed (see lshosts(1) to find 
out what resources are available on each host).

.IP
If one or more host names are specified, only load information about 
the hosts that satisfy the resource requirements is displayed.


.TP 
\fIhost_name\fR ... 

.IP
Displays only load information for the specified hosts. 


.TP 
-s [\fIresource_name\fR ...]

.IP
Displays information for all dynamic shared resources configured in the 
cluster.

.IP
If resources are specified, only displays information for the specified 
resources. \fIresource_name\fR must be a dynamic shared resource name.

.IP



.TP 
\fB-h
\fR
.IP
Prints command usage to stderr and exits.


.TP 
\fB-V
\fR
.IP
Prints openlava release version to stderr and exits.


.SH OUTPUT
.BR
.PP
.SS HOST-BASED OUTPUT (Default output)
.BR
.PP
.PP
Numeric dynamic non-shared resources are displayed. The selection 
and order sections of \fIres_req\fR control for which hosts information is 
displayed and how they are ordered (see lsfintro(1)).
.PP
The displayed default load information includes the following fields:

.IP
HOST_NAME
.BR
.RS
.IP
Standard host name used by openlava, typically an Internet domain name 
with two components.

.RE

.IP
status
.BR
.RS
.IP
Status of the host. A minus sign (-) may precede the status, 
indicating that the Remote Execution Server (RES) on the host is not 
running.

.IP
Possible statuses are:


.IP
ok
.BR
.RS
.IP
The host is in normal load sharing state and can accept remote 
jobs.

.RE

.IP
busy
.BR
.RS
.IP
The host is overloaded because some load indices exceed 
configured thresholds. Load index values that caused the host 
to be busy are preceded by an asterisk (*). Built-in load indices 
include r15s, r1m, r15m, ut, pg, io, ls, it, swp, mem and tmp 
(see below). External load indices are configured in the file 
lsf.cluster.\fIcluster_name\fR (see lsf.cluster(5)).


.IP
r15s
.BR
.RS
.IP
The 15-second exponentially averaged CPU run queue 
length.

.RE

.IP
r1m
.BR
.RS
.IP
The 1-minute exponentially averaged CPU run queue 
length.

.RE

.IP
r15m
.BR
.RS
.IP
The 15-minute exponentially averaged CPU run queue 
length.

.RE

.IP
ut
.BR
.RS
.IP
The CPU utilization exponentially averaged over the last 
minute, between 0 and 1.

.RE

.IP
pg
.BR
.RS
.IP
The memory paging rate exponentially averaged over the 
last minute, in pages per second.

.RE

.IP
io
.BR
.RS
.IP
The disk I/O rate exponentially averaged over the last 
minute, in KB per second (this is only available when the -l 
option is specified).

.RE

.IP
ls
.BR
.RS
.IP
The number of current login users.

.RE

.IP
it
.BR
.RS
.IP
On UNIX, the idle time of the host (keyboard not touched 
on all logged in sessions), in minutes.

.RE

.IP
swp
.BR
.RS
.IP
The amount of swap space available, in megabytes.

.RE

.IP
mem
.BR
.RS
.IP
The amount of available memory, in megabytes.

.RE

.IP
tmp
.BR
.RS
.IP
The amount of free space in /tmp, in megabytes.

.RE

.IP
external_index
.BR
.RS
.IP
Any site-configured global external load indices (see 
lim(8)). Available only when the -l option or the -I 
option with the index name is used, and only if defined in 
the lsf.cluster.\fIcluster_name\fR (see lsf.cluster(5)) 
configuration file. Note that \fIexternal_index\fR should not 
contain shared resources.

.RE
.RE

.IP
lockW
.BR
.RS
.IP
The host is locked by its run window. Run windows for a host 
are specified in the configuration file (see lsf.conf(5)) and 
can be displayed by lshosts. A locked host will not accept 
load shared jobs from other hosts.

.RE

.IP
lockU
.BR
.RS
.IP
The host is locked by the openlava administrator or root.

.RE

.IP
unavail
.BR
.RS
.IP
The host is down or the Load Information Manager (LIM) on the 
host is not running.

.RE
.RE
.RE
.SS RESOURCE-BASED OUTPUT (lsload -s )
.BR
.PP
.PP
Displays information about dynamic shared resources. Each line gives 
the value and the associated hosts for an instance of the resource. See 
lim(8), and lsf.cluster(5) for information on configuring 
dynamic shared resources.
.PP
The displayed information consists of the following fields:

.IP
RESOURCE
.BR
.RS
.IP
Name of the resource.

.RE

.IP
VALUE
.BR
.RS
.IP
Value for an instance of the resource.

.RE

.IP
LOCATION
.BR
.RS
.IP
Hosts associated with the instance of the resource.

.RE
.SH EXAMPLES
.BR
.PP
.PP
% \fBlsload -R "select[r1m<=0.5 && swp>=20 && type==ALPHA]"\fR
.PP
OR, in restricted format:
.PP
%\fB \fR\fBlsload -R r1m=0.5:swp=20:type=ALPHA\fR
.PP
Displays the load of ALPHA hosts with at least 20 megabytes of swap 
space, and a 1-minute run queue length less than 0.5.
.PP
% \fBlsload -R "select[(1-swp/maxswp)<0.75] order[pg]"\fR
.PP
Displays the load of the hosts whose swap space utilization is less than 
75%. The resulting hosts are ordered by paging rate.
.PP
%\fB \fR\fBlsload -I r1m:ut:io:pg\fR
.PP
Displays the 1-minute CPU raw run queue length, the CPU utilization, 
the disk I/O and paging rates for all hosts in the cluster.
.PP
%\fB \fR\fBlsload -E\fR
.PP
Displays the load of all hosts, ordered by r15s:pg\fB,\fR with the CPU run 
queue lengths being the effective run queue lengths (see 
lsfintro(1)).
.PP
%\fB \fR\fBlsload -s verilog_license\fR
.PP
Displays the value and location of all the verilog_license dynamic 
shared resource instances.
.SH SEE ALSO
.BR
.PP
.PP
lsfintro(1), lim(8), lsf.cluster(5), lsplace(1), 
lshosts(1), lsinfo(1), lslockhost(8), ls_load(3)
.SH DIAGNOSTICS
.BR
.PP
.PP
Exit status is -10 if an openlava problem is detected or a bad resource name 
is specified.
.PP
Exit status is -1 if a bad parameter is specified, otherwise lsload 
returns 0.
.PP


NEWS
Release notes for OpenLava versions

3.3.pre.2
=========
-- Keep the lsb.events and lsb.acct files always open to improve
   file access performance.
-- Implement the switching of the lsb.events asynchronously in mbatchd
   child process.
-- Implemented the bjdep command. bjdep jobid prints the status of
   job dependencies for the given job.
-- Add the bjdep man page.
-- Implement NIOS_RWAIT_SELECT, if configure in lsf.conf then nios
   and ls_rwait() will not use the SIGUSR1 to detect a task that
   has finished. ls_rwait() will select() on a connection with nios
   instead and nios will send a message when a task exits.
-- Fix RES NIOS protocol bug that prevented multiple tasks running on
   the same host from being collected using ls_rwait().
-- Implemented job groups. New commands are bgadd, bgdel and bgroups.
-- Implemented bjobs -UF unformatted bjobs output.
-- Add LSF_BINDIR in openlava startup scripts.
-- Implement bjobs -g to display only selected job groups.
-- Implement bkill -g to kill only jobs in a specific job group.

3.3.pre.1
=========
-- Fix fairshare algorithm to decrease the number of jobs per user
   only if the job has been started.
-- Fix a bug in ownership policy where wrong runtime comparison was
   performed.
-- Implement the replay of job absolute runtime for ownership policy.
-- Implement floating client feature. If the user configures in lsf.conf
   LIM_ACCEPT_FLOAT_CLIENT=y any host can submit jobs and query the
   daemons without being registered in the lsf.cluster file.

3.2.1
======
-- Fix the parsing algorithm for compact host name.

3.2.0
======
-- Fix main Makefile.am to include the m4 directory.

3.2.beta1
=========
-- Hosts member od a host group can now be specified in a compact format
   for example: (ol[1-3]).
-- Implement configure option --enable-debug. If set the compilation will
   have the -Werror option enabled. By default this option is not set.

3.2.pre
========
-- badmin ckconfig/reconfig displays configuration error and warning
   messages.
-- bsub/bmod/brestart -h displays usage information.
-- Added USER_SHARES information in lsb.users.5
-- Add lsclusters command.
-- Use qsort() to sort jobs in MBD this improves the performance
   of replay and general bsub operations.
-- Introduce a getc2() functions to read lsb.events in memory
   before parsing it.
-- Reduce the number of malloc operations when reading the
   lsb.events file.
-- Implement memory reservation if the queue has SLOT_MEMORY_RESERVE configured
   and the job requests rusage of memory. The jobs has to reserve slots
   on the hosts as well.
-- Implement a correct switching of the lim.events file.
-- Update the bmgroup command to display GROUP_SLOT and MAX_SLOTS
   sections if configured.
-- Update the bqueues command to display slot and memory reservation.
-- Allow batch hosts configured in lsb.hosts to be specified in a
   compact form, i.e. ol[1-3] the mbatchd will expand example this into 3
   hosts ol1, ol2, ol3.
-- Allow OpenLava to use cores as cpus regardless is the machines has
   hyperthreads or not. To enable this behaviour set the variable
   LIM_DEFINE_NCPUS=cores|threads in lsf.conf. Default is threads.
-- Fix bsub bug in which if user specify -n min,max the max was ignored.
-- Make fairshare configuration compatible with the LSF one.

3.1.4
======

-- Fix the problem that if the order of events in lim.events is not correct
   LIM may core dump.
-- Change the ownership of the mbatchd log file to avoid messages going
   to syslog messages file.
-- Prevent LIM from core dumping in findHost() if the input name is NULL.
-- Enable the secondary admin to run 'badmin reconfig'.
-- Fix res nios protocol to pass the correct taskid to nios.

3.1.3
======
-- Corrected the lsf.cluster man page describing LIM communication
   parameters.
-- Enable LIM communication debugging using LC_COMM class and
   LOG_INFO level.
-- Fix bug in announceMaster() which was slowing down announcement
   of slave hosts.
-- Fix a bug in the slave host timer causing slave to send load
   to master more frequently then necessary.
-- Fix counter update in fairshare policy when jobs are suspended
   and resumed.
-- Remove spurious res log messages.

3.1.2
=====
-- Fix a critical mbatchd bug when sorting queues.
-- Enable the openlava.spec file and the rpm.sh script
   which builds the rpm.

3.1.1
=====
-- Fix scheduling bug in which performance could drop if the lowest
   priority queue was always full.
-- Fix the incorrect encoding of bchkpnt command data structure.

3.1.0
======
-- Modify openlava.sh to forcefully terminate sbatchd if SIGTERM does
   not work.
-- Fix LIM bug in which the code was accessing non initialized memory.
-- Fix several memory leaks in LIM.
-- Fix res hang in child handler.
-- Fix memory corruption in mbatchd when dealing with rusage.
-- Fix mbatchd child core dump when processing bugroup command.
-- Improve fairshare performance.
-- Introduce an lsf.conf parameter MBD_MAX_JOB_SCHED which allows site
   to configure how many jobs should MBD attempt to schedule per cycle.
   Typically two times the number of slots.
-- Introduce an lsf.conf parameter MBD_QSORT_JOBS which changes the way
   jobs are inserted and sorted in the PJL. This parameter avoids searching
   the job list upon a new job arrival speeding up the submission and overall
   MBD operations.
-- Fix issues in the job sorting order while using qsort().
-- Increase the array size of LIM clients from 64 to 2014.
-- Fix lim core dump if the lim receives an operation request from
   outside of the OL cluster and it is not LIM_ADD_HOST.
-- Fix bsub < file which was not creating the shell file correctly.
-- Fix rusage bug in which the host could be allocated to a job even with
   not enough resources.
-- Fix memory error in mbatchd when computing job's rusage.
-- Fix various memory leaks

3.1pre1
=======

-- Removed obsolete mls code and infrastructure simplifying user
   management.

lsf/lstools/lsload.c
/* $Id: lsload.c 397 2007-11-26 19:04:00Z mblack $
 * Copyright (C) 2007 Platform Computing Inc
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of version 2 of the GNU General Public License as
 * published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.

 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA
 *
 */


#include "../lsf.h"
#include "../lib/lib.table.h"
#include "../lib/lproto.h"
#include "../intlib/intlibout.h"

#define MAXLISTSIZE 256

extern int makeFields(struct hostLoad *, char *loadval[], char **);
extern int makewideFields(struct hostLoad *, char *loadval[], char **);
extern char *formatHeader(char **, char);
extern char *wideformatHeader(char **, char);
extern char **filterToNames(char *);
extern int num_loadindex;

void
usage(char *cmd)
{
    fprintf(stderr,"Usage:\n%s [-h] [-V] [-N|-E] [-l] [-R res_req] [-I index_list] [-n num_hosts] [host_name ... | cluster_name ...]\n", cmd);
   fprintf(stderr, "or\n%s [-h] [-V] -s [ shared_resource_name ... ]\n", cmd);
}

int
main(int argc, char **argv)
{
    int i,j, num, numneeded;
    char *resreq = NULL;
    struct hostLoad *hosts;
    char *hostnames[MAXLISTSIZE];
    char statusbuf[20];
    int options = 0;
    static char **loadval;
    char *indexfilter = NULL;
    char **nlp;
    static char *defaultindex[] = {"r15s", "r1m", "r15m", "ut", "pg", "ls",
                                   "it", "tmp", "swp", "mem", NULL};
    int	achar;
    char longFormat = FALSE;
    char wideFormat = FALSE;
    char badHost = FALSE, sOption = FALSE, otherOption = FALSE;
    int extView = FALSE;
    char **shareNames, **shareValues, **formats;
    int isClus, retVal = 0;

    num = 0;
    numneeded = 0;
    opterr = 0;

    if (ls_initdebug(argv[0]) < 0) {
        ls_perror("ls_initdebug");
        exit(-1);
    }

    for (i = 1; i < argc; i++) {
       if (strcmp(argv[i], "-h") == 0) {
           usage(argv[0]);
           exit (0);
       } else if (strcmp(argv[i], "-V") == 0) {
           fputs(_LS_VERSION_, stderr);
           exit(0);
       } else if (strcmp(argv[i], "-s") == 0) {
           if (otherOption == TRUE) {
               usage(argv[0]);
               exit(-1);
           }
           sOption = TRUE;
           optind = i + 1;
       } else if (strcmp(argv[i], "-e") == 0) {
           if (otherOption == TRUE || sOption == FALSE) {
               usage(argv[0]);
               exit(-1);
           }
           extView = TRUE;
           optind = i + 1;
       } else if (strcmp(argv[i], "-R") == 0 || strcmp(argv[i], "-l") == 0
                  || strcmp(argv[i], "-I") == 0 || strcmp(argv[i], "-N") == 0
                  || strcmp(argv[i], "-E") == 0 || strcmp(argv[i], "-n") == 0
		  || strcmp(argv[i], "-w") == 0 ) {
            otherOption = TRUE;
            if (sOption == TRUE) {
                usage(argv[0]);
                exit(-1);
            }
        }
    }

    if (sOption == TRUE) {
        displayShareResource(argc, argv, optind, FALSE, extView);
        return 0;
    }

    while ((achar = getopt(argc, argv, "R:I:NEln:w")) != EOF) {
	switch (achar) {
	case 'R':
            resreq = optarg;
            break;
	case 'I':
            indexfilter = optarg;
            break;

	case 'N':
	    if (options & EFFECTIVE)
		usage(argv[0]);
            options = NORMALIZE;
	    break;

        case 'E':
	    if (options & NORMALIZE)
		usage(argv[0]);
             options = EFFECTIVE;
	     break;
	case 'n':
            numneeded = atoi(optarg);
            if (numneeded <= 0)
                usage(argv[0]);
            break;

        case 'l':
	    longFormat = TRUE;
	    if (wideFormat == TRUE)
		usage(argv[0]);
	    break;

	case 'w':
	    wideFormat = TRUE;
	    if (longFormat == TRUE)
		usage(argv[0]);
	    break;

	case 'V':
	    fputs(_LS_VERSION_, stderr);
	    exit(0);

	case 'h':
	default:
            usage(argv[0]);

        }
    }

    for ( ; optind < argc ; optind++) {
        if (num >= MAXLISTSIZE) {
            fprintf(stderr, "too many hosts specified (maximum %d)\n",
                    MAXLISTSIZE);
            exit(-1);
        }
        if ( (isClus = ls_isclustername(argv[optind])) < 0 ) {
	    fprintf(stderr, "lsload: %s\n", ls_sysmsg());
            badHost = TRUE;
	    continue;
	} else if ( (isClus == 0) &&
		    (!Gethostbyname_(argv[optind])) ) {
	    fprintf(stderr, "\
%s: invalid hostname %s\n", __func__, argv[optind]);
            badHost = TRUE;
            continue;
	}
        hostnames[num] = argv[optind];
        num++;
    }

    if (num == 0 && badHost)
        exit(-1);

    if (!longFormat) {
        if (indexfilter)
            nlp = filterToNames(indexfilter);
        else
            nlp = defaultindex;
    } else {
        nlp = NULL;
    }

    TIMEIT(0, (hosts = ls_loadinfo(resreq,
                                   &numneeded,
                                   options,
                                   0,
                                   hostnames,
                                   num,
                                   &nlp)), "ls_loadinfo");
    if (!hosts) {
	ls_perror("lsload");
        exit(-10);
    }

    if (longFormat)
        printf("%s", formatHeader(nlp, longFormat));
    else
	if (wideFormat)
	    printf("%s\n", wideformatHeader(nlp, longFormat));
        else
            printf("%s\n", formatHeader(nlp, longFormat));

    if (!(loadval = malloc(num_loadindex * sizeof(char *)))) {
        lserrno = LSE_MALLOC;
        ls_perror("lsload");
        exit(-1);
    }

    for (i = 0;i < numneeded; i++) {
	if (LS_ISUNAVAIL(hosts[i].status))
	    strcpy(statusbuf, "unavail");
	else {
	    statusbuf[0] = '\0';
            if (LS_ISRESDOWN(hosts[i].status))
		strcat(statusbuf, "-");
	    if (LS_ISOKNRES(hosts[i].status))
		strcat(statusbuf, "ok");
	    if (LS_ISLOCKEDU(hosts[i].status) && !LS_ISLOCKEDW(hosts[i].status))
		strcat(statusbuf, "lockU");
	    if (LS_ISLOCKEDW(hosts[i].status) && !LS_ISLOCKEDU(hosts[i].status))
		strcat(statusbuf, "lockW");
	    if (LS_ISLOCKEDW(hosts[i].status) && LS_ISLOCKEDU(hosts[i].status))
		strcat(statusbuf, "lockUW");
	    if (LS_ISBUSY(hosts[i].status) && !LS_ISLOCKED(hosts[i].status))
		strcat(statusbuf, "busy");
	}

        if (longFormat) {
            retVal = makeShareField(hosts[i].hostName, FALSE, &shareNames,
                                    &shareValues, &formats);
	    if (i == 0) {

		if (retVal > 0) {

		    for (j = 0; j < retVal; j++) {
			printf(formats[j], shareNames[j]);
		    }
		}
		putchar('\n');
	    }
	    printf("%-23s %6s", hosts[i].hostName, statusbuf);
        } else
	    printf("%-15.15s %6s", hosts[i].hostName, statusbuf);

	if (!LS_ISUNAVAIL(hosts[i].status)) {
            int nf;
	    if (wideFormat){
		nf = makewideFields(&hosts[i], loadval, nlp);
	    }
	    else
	    nf = makeFields(&hosts[i], loadval, nlp);
            for(j = 0; j < nf; j++)
                printf("%s",loadval[j]);
            if (retVal > 0) {
                for (j = 0; j < retVal; j++) {
                    printf(formats[j], shareValues[j]);
                }
            }
	}
	putchar('\n');
    }

    exit (0);
}
